### PR TITLE
L66: Core server RPCs will not report cancellation if completed with non-OK status

### PR DESCRIPTION
**Pull request by vjpai**
This is a breaking API change at core but not at the wrapped languages.

This is implemented in grpc/grpc#22991

### CODE REVIEW

File: L66-core-cancellation-status.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+Calling an RPC cancelled if it completes for entirely expected reasons is confusing. Generally speaking, this distinction has not been noticed at the API layers but will become more common once the C++ callback API becomes formalized (since that has an OnCancel reaction for server RPCs). As a result, this issue should be clarified and fixed before the C++ callback API becomes common.
+
+## Implementation
+
+1. Decide if a server call is canceled by whether or not it could successfully send status
+1. Fix core tests (wrapped language tests already had the proper expectations)
+1. Mark a call failed according to channelz if it was canceled or if the server sent a non-OK status
+1. Add a C++ test to validate that OnCancel is not called on non-OK explicit status
+ 1. Fix the core surface comment about this topic and bump the core major API version
```
Comment (chwarr) [inline L66-core-cancellation-status.md:44]: Minor: this is indented by one space. For a sublist, I think you need at least three spaces, but four is common.

File: L66-core-cancellation-status.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+## Open issues (if applicable)
+
+* Does this affect the core wrapped language APIs?
+  - This distinction is not visible to Ruby or the C++ synchronous API at all since the RPC in both of those cases is complete when returning status. Thus cancellation checks just see the effect of other forms of cancellation.
+  - For the C++ asynchronous API, `IsCancelled` currently doesn't define its result. This change will also include an explciit definition of cancellation for C++.
+  - Python does not expose this functionality to the server application.
+  - For C#, ???
+
+* Should this only trigger on explicit cancellations?
+- In practice, it doesn't matter if the application requested cancellation or cancellation happened because of a deadline or network problem; in any case, the RPC is failed for reasons out of the scope of the RPC. In practice, issues like deadline exceeded could be implemented using cancellation, so there is no strong distinction in any case.
```
Comment (chwarr) [inline L66-core-cancellation-status.md:55]: Minor: I think the lines starting with `-` are intended to be underneath each item starting with `*`. To get that to render, they need to be indented. (Minimum three spaces, IIRC, but four is common.)

File: L66-core-cancellation-status.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+## Background
+
+gRPC core versions up to 10.X specify that the out argument of the `GRPC_OP_RECV_CLOSE_ON_SERVER` operation is a pointer to an int that specifies whether the RPC was cancelled. The definition specified in the comment is
+
+```
+      /** out argument, set to 1 if the call failed in any way (seen as a
+          cancellation on the server), or 0 if the call succeeded */
+```
+
+The term `in any way` is not clear, and the broadest (and current) definition is that it accounts for failures caused by client-side or server-side cancellations, deadline exceeded, maximum connection age exceeded, network resets, *and also explicit sending of non-OK status*.
```
Comment (chwarr) [inline L66-core-cancellation-status.md:23]: Minor:

```suggestion
The phrase "in any way" is not clear, and the broadest (and current) definition is that it accounts for failures caused by client-side or server-side cancellations, deadline exceeded, maximum connection age exceeded, network resets, *and also explicit sending of non-OK status*.
```

Comment (chwarr) [inline L66-core-cancellation-status.md:23]: Minor:

```suggestion
The term `in any way` is not clear, and the broadest (and current) definition is that it accounts for failures caused by client-side or server-side cancellations, deadlines being exceeded, maximum connection age being exceeded, network resets, *and also explicit sending of non-OK status*.
```

I couldn't think of a better way to "noun" "exceeded"... "Expirations"? But that doesn't really work for max connection age...

File: L66-core-cancellation-status.md (hunk 1 above)
Comment (vjpai) [reply L66-core-cancellation-status.md]: Fixed. Thanks1

File: L66-core-cancellation-status.md (hunk 3 above)
Comment (vjpai) [reply L66-core-cancellation-status.md]: connections exceeding their maximum age ?

Comment (chwarr) [reply L66-core-cancellation-status.md]: SGTM

File: L66-core-cancellation-status.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+```
+      /** out argument, set to 1 if the call failed in any way (seen as a
+          cancellation on the server), or 0 if the call succeeded */
+```
+
+The phrase "in any way" is not clear, and the broadest (and current) definition is that it accounts for failures caused by issues such as client-side or server-side cancellations, deadlines being exceeded, connections exceeding their maximum age, network resets, *and explicit sending of non-OK status*.
+
+
+### Related Proposals
```
Comment (markdroth) [inline L66-core-cancellation-status.md:26]: This can point to the gRFC for the callback API.

File: L66-core-cancellation-status.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+1. Fix the core surface comment about this topic and bump the core major API version
+1. Add a C++ comment clarifying the meaning of `IsCancelled`
+
+## Open issues (if applicable)
+
+* Does this affect the core wrapped language APIs?
+  - This distinction is not visible to Ruby or the C++ synchronous API at all since the RPC in both of those cases is complete when returning status. Thus cancellation checks just see the effect of other forms of cancellation.
+  - For the C++ asynchronous API, `IsCancelled` currently doesn't define its result. This change also includes an explciit definition of cancellation for C++.
+  - Python does not expose this functionality to the server application.
+  - For C#, ???
```
Comment (markdroth) [inline L66-core-cancellation-status.md:55]: Please consult with @jtattermusch about this.  We should make sure this entry is filled in before merging this or the implementation PR.

File: L66-core-cancellation-status.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+1. Decide if a server call is canceled by whether or not it could successfully send status
+1. Fix core tests (wrapped language tests already had the proper expectations)
+1. Mark a call failed according to channelz if it was canceled or if the server sent a non-OK status
+1. Add a C++ test to validate that OnCancel is not called on non-OK explicit status
+1. Fix the core surface comment about this topic and bump the core major API version
+1. Add a C++ comment clarifying the meaning of `IsCancelled`
+
+## Open issues (if applicable)
+
+* Does this affect the core wrapped language APIs?
```
Comment (markdroth) [inline L66-core-cancellation-status.md:51]: What about PHP and ObjC?

File: L66-core-cancellation-status.md (hunk 4 above)
Comment (vjpai) [reply L66-core-cancellation-status.md:26]: Done.

File: L66-core-cancellation-status.md (hunk 6 above)
Comment (vjpai) [reply L66-core-cancellation-status.md:51]: Not server languages, so this doesn't apply to them.

File: L66-core-cancellation-status.md (hunk 5 above)
Comment (vjpai) [reply L66-core-cancellation-status.md:55]: I've reached out again to him. I'll let you know when I fill in this content.

**Comment by markdroth** [comment]
This looks great to me.  Just waiting for the C# info before approving.

**Comment by vjpai** [comment]
@jtattermusch Thanks for the discussion this morning/evening, can you confirm that you're satisfied with the content in e677f57 ?

**Comment by jtattermusch** [comment]
> @jtattermusch Thanks for the discussion this morning/evening, can you confirm that you're satisfied with the content in [e677f57](https://github.com/grpc/proposal/commit/e677f5701cc5816da10172a16fe9f105140702fe) ?

The C# specific explanation seems fair (LGTM on that part)

@JamesNK I think for grpc-dotnet this is basically the same story, can you please doublecheck the grpc-dotnet behavior of ServerCallContext.CancellationToken?

**Comment by vjpai** [comment]
Barring any further discussion, this will be ready for merger on June 11, followed shortly by grpc/grpc#22991

**Pull request merged**